enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,210 @@
+see-also:
+  - microshift-apiserver-certs.md
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# MicroShift API server custom external certificates
+## Summary
+
+This enhancement extends the Microshift apiserver Certs to allow the user to
+provide additional custom certificates (e.g. signed by a 3rd party CA) for API server SSL handkshake and external authentication.
+
+
+> Anytime the document mentions API server, it refers to kube API server.
+
+
+## Motivation
+Currently, users of Microshift cannot provide their own generated certificates,
+Some customers have very strict requirements regarding TLS certs. There are frequently intermediate proxies which allow only TLS traffic with certs that are signed by organization owned CAs.
+
+### User Stories
+* As a MicroShift administrator, I want to be able to add  organization generated certificates. 
+  - each certificate may contain:
+    - Single Common Name containing The API server DNS/IPAddress.
+    - Multiple Subject Alternative Names (SAN) containing the API server DNS/IPAddress.
+
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate.
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+N/A
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file.
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+because we dont want to disrupt the internal API communication and make sure the internal certificate will be automaticly renewed,
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+The generated `kubeconfig` for the custom certificates will omit the certificate-authority-data section,
+therefore custom certificates will have to be validated against CAs in the RHEL trust store. 
+
+For every certificate that contain `names` the first non-wildcard `name`  will be used,
+and placed inside the `server` section of the generated `kubeconfig`.
```

Comment:
I might be missing something here about certs, but what happens when `names` are given and certificates has its own name/SANs - are the lists merged with non-wildcard names being on the top?

Reply:
so we can  use CN as unique identifier for the `server` section of generated kubeconfig.
in such case kubeconfigs wont be generated for wild-card certificates.

Reply:
It seems like if the certificate has a wildcard but the user gives us a more specific name, we could generate the kubeconfig for that name. That's similar to what happens with the certs we generate, if they give us a SAN, right?

Reply:
currently if the user give us a SAN ,then we generate external Certs with that SAN.

I've made more experiments with names,  configured  a `name`  (ie: nameonly.com ) that is not  in the Certificate CN  (nor contained in the wildcard), 
and  placed that `name` in kubeconfig under `server` :
`server: https://nameonly.com:6443/`

what happen is that the  oc client rejected that `server` (since it doesnt appear in the actual certificate)
```
Unable to connect to the server: tls: failed to verify certificate: x509: certificate is valid for api.example.com, localhost.localdomain, 192.168.2.130, not nameonly.com
```

so names are only useful if combined with wildcards inside the cert. (wildcard certs)
for cases when the Certificate is created for multiple nodes and there is need  to specify that specific node node.

Reply:
That makes sense, and that's the case I was trying to describe. When the cert has `*.example.com` and the config file has `my-host.example.com`, we should generate a kubeconfig with a server URL pointing to `my-host.example.com`.  If the user gives us only the cert with the wildcard `*.example.com` then I think we said the other day in the team meeting that we would generate a kubeconfig with the server URL using the FQDN of the host.

Reply:
@dhellmann , see my [0]  commit , i've added the meeting outcome there [0] https://github.com/openshift/enhancements/pull/1541/commits/1c92195c26040867f8a9c537557114c5a7eb682b